diff/diff.go

```diff
@@ -0,0 +1,534 @@
+	wg               sync.WaitGroup
+}
+
+// NewDiff returns a Diff instance.
+func NewDiff(db1, db2 *sql.DB, cfg *Config) (diff *Diff, err error) {
+	diff = &Diff{
+		db1:              db1,
+		db2:              db2,
+		dbName:           cfg.SourceDBCfg.Name,
+		chunkSize:        cfg.ChunkSize,
+		sample:           cfg.Sample,
+		checkThreadCount: cfg.CheckThreadCount,
+		useRowID:         cfg.UseRowID,
+		tables:           cfg.Tables,
+		sqlCh:            make(chan string),
+	}
+	for _, table := range diff.tables {
+		table.Info, err = pkgdb.GetTableInfoWithRowID(diff.db1, diff.dbName, table.Name, cfg.UseRowID)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		table.Schema = diff.dbName
+	}
+	diff.fixSQLFile, err = os.Create(cfg.FixSQLFile)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return diff, nil
+}
+
+// Equal tests whether two database have same data and schema.
+func (df *Diff) Equal() (equal bool, err error) {
+	defer func() {
+		df.fixSQLFile.Close()
+		df.db1.Close()
+		df.db2.Close()
+	}()
+
+	df.wg.Add(1)
+	go df.WriteSqls()
+
+	equal = true
+	tbls1, err := pkgdb.GetTables(df.db1, df.dbName)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	tbls2, err := pkgdb.GetTables(df.db2, df.dbName)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	eq := equalStrings(tbls1, tbls2)
+	// len(df.tables) == 0 means check all tables
+	if !eq && len(df.tables) == 0 {
+		log.Errorf("show tables get different table. [source db tables] %v [target db tables] %v", tbls1, tbls2)
+		equal = false
+	}
+
+	if len(df.tables) == 0 {
+		df.tables = make([]*TableCheckCfg, 0, len(tbls1))
+		for _, name := range tbls1 {
+			table := &TableCheckCfg{Name: name, Schema: df.dbName}
+			table.Info, err = pkgdb.GetTableInfoWithRowID(df.db1, df.dbName, name, df.useRowID)
+			if err != nil {
+				return false, errors.Trace(err)
+			}
+
+			df.tables = append(df.tables, table)
+		}
+	}
+
+	for _, table := range df.tables {
+		tableInfo1 := table.Info
+		tableInfo2, err := pkgdb.GetTableInfoWithRowID(df.db2, df.dbName, table.Name, df.useRowID)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		eq, err = df.EqualTableStruct(tableInfo1, tableInfo2)
+		if err != nil {
+			return equal, errors.Trace(err)
+		}
+		if !eq {
+			log.Errorf("table have different index: %s\n", table.Name)
+			equal = false
+		}
+
+		eq, err = df.EqualTableData(table)
+		if err != nil {
+			log.Errorf("equal table error %v", err)
+		}
+		if !eq {
+			equal = false
+			log.Errorf("table %s's data is not equal", table.Name)
+		}
+	}
+
+	df.sqlCh <- "end"
+	df.wg.Wait()
+	return
+}
+
+// EqualTableStruct tests whether two table's struct are same.
+func (df *Diff) EqualTableStruct(tableInfo1, tableInfo2 *model.TableInfo) (bool, error) {
+	// check columns
+	if len(tableInfo1.Columns) != len(tableInfo2.Columns) {
+		return false, nil
+	}
+
+	for j, col := range tableInfo1.Columns {
+		if col.Name.O != tableInfo2.Columns[j].Name.O {
```

Comment:
need support difference order with tb1 and tb2?

Reply:
we can make it more flexible later